### ISSUE TITLE
Gaufrette orphanage

### ISSUE THREAD

**Issue by Looted** [1up-lab/OneupUploaderBundle#159]
Hello!
I have a question regarding orphanage with the gaufrette filesystem. I need to use orphanage along with Amazon S3. I need to implement removing items from orphanage however, so files can be dynamically removed from dropzone (so that if someone uploads two files, removes one and submits the form, only one gets moved from the orphanage). Right now the config is as follows:

```
oneup_uploader:
    mappings:
        job_pdf:
            frontend: dropzone
            storage:
                type: gaufrette
                filesystem: gaufrette.test_filesystem
            use_orphanage: true
    orphanage:
        maxage: 86400
    directory: orphanage
```

The filesystem points to web/uploads for now (it will be switched to Amazon S3 on production server). The orphanage saves files outside the filesystem however - not relative to the filesystem root. It should be web/uploads/orphanage, but saves to web/orphanage. Please advise. 

With regards

**Comment by Looted** [comment]
The problem seems to be realted to gaufrette - whenever I try to save files to directories on a gaufrette fs, the directories are ignored. For example with config:

    oneup_uploader:
    mappings:
        documents:
            frontend: dropzone
            storage:
                type: gaufrette
                filesystem: gaufrette.test_filesystem
                stream_wrapper: ~
                directory: documents

The files get saved to the base path, and not the documents directory. It is the same for gaufrette local storage or AWS S3.

**Comment by Thommas** [comment]
I'm using this, maybe that can help you :

    services:
        myproject.aws_s3.client:
            class: AmazonS3
            arguments:
                options:
                    key: %amazon_s3_key%
                    secret: %amazon_s3_secret%
      myproject.aws_s3.adapter.myentity:
            class: Gaufrette\Adapter\AmazonS3
            arguments: [@myproject.aws_s3.client, %amazon_s3_bucket_name%, {'directory': 'mydirectory'}]
        myproject.aws_s3.filesystem.myentity:
            class: Gaufrette\Filesystem
            arguments: [@usn_accounting.aws_s3.adapter.myentity]

    knp_gaufrette:
        adapters:
            myadaptater:
                amazon_s3:
                    amazon_s3_id: myproject.aws_s3.client
                    bucket_name: %amazon_s3_bucket_name%
        filesystems:
            myfs:
                adapter: myadaptater

    oneup_uploader:
        mappings:
            mymapping:
                frontend: dropzone
                storage:
                    type: gaufrette
                    filesystem: myproject.aws_s3.filesystem.myfs

**Comment by naabys** [comment]
@Thommas are you positive on the fact that orphanage is directed to s3 storage with that configuration ? Because as far as i'm concerned, having a similar configuration, my orphan files are uploaded to cache directory. I don't understand

**Comment by Thommas** [comment]
The configuration I showed you was without orphanage enabled.

To enable orphanage use this.
    oneup_uploader:
        mappings:
            myentitymapping:
                use_orphanage: true
                frontend: dropzone
                storage:
                    type: gaufrette
                    filesystem: mycompany.aws_s3.filesystem.myentity
        orphanage:
            directory: %kernel.root_dir%/../web/uploads

If you do this, remember you need to call in your controller where you save your entity:
            $oneupUploaderManager = $this->get('oneup_uploader.orphanage_manager')->get('myentitymapping');
            $oneupUploaderManager->uploadFiles();

First your file is uploaded to the orphanage local directory, then when you call the code in the controller, the file is uploaded to AmazonS3.

Tell me if you need any help to achieve the right configuration.

**Comment by Thommas** [comment]
oneup_uploader:
        mappings:
            myentitymapping:
                use_orphanage: true
                frontend: dropzone
                storage:
                    type: gaufrette
                    filesystem: mycompany.aws_s3.filesystem.myentity
        orphanage:
            directory: %kernel.root_dir%/../web/uploads

Controller:

    $oneupUploaderManager = $this->get('oneup_uploader.orphanage_manager')->get('myentitymapping');
    $oneupUploaderManager->uploadFiles();

**Comment by Thommas** [comment]
"The files get saved to the base path, and not the documents directory. It is the same for gaufrette local storage or AWS S3."
If you read the documentation, they explained why on the local directory, the path follow a static pattern with the user SessionID. If you're trying to change how files are stored on the local directory when using orphanage, I believe this is just not possible for now.
However, if you follow my configuration, the directory will be used when uploading to AmazonS3.
It's working for me.

    "oneup/uploader-bundle": "~1.3",
    "knplabs/knp-gaufrette-bundle": "*@dev",

**Comment by naabys** [comment]
As a matter of fact, I got the orphanage to work. 
I thought it was somehow possible, and straightforward via configuration to use S3 as an orphanage.
Here's my use case:
Once they are approved by moderator, my files are uploaded from orphanage to S3.
I plan to deploy the application I'm working on to a AWS EBS environment. But the local storage is attached to EC2 instances and can be wiped any time (auto-scaling, app update...). 
In order to be the more 12-factors compliant as possible, I needed to be able to configure a s3 bucket as orphanage. 
Now I think i'll go with an NFS storage.

Thanks for your insight, anyway!
